plugins/modules/gpg_keypair.py

```diff
@@ -0,0 +1,507 @@
+            module.fail_json('Invalid usage for {0} {1}.'.format(key_type, key_name))
+        pass
+    elif key_type == 'DSA':
+        if key_usage and tuple(key_usage) not in all_permutations(['sign', 'auth', 'cert']):
+            module.fail_json('Invalid usage for {0} {1}.'.format(key_type, key_name))
+        pass
+    elif key_type == 'ELG':
+        if key_name == 'primary key':
+            module.fail_json('Invalid type for {0}.'.format(key_name))
+        elif key_usage and key_usage != ['encr']:
+            module.fail_json('Invalid usage for {0} {1}.'.format(key_type, key_name))
+        pass
+
+
+def delete_keypair(module, matching_keys, check_mode):
+    if matching_keys:
+        module.run_command(
+            [
+                '--dry-run' if check_mode else '',
+                '--batch',
+                '--yes',
+                '--delete-secret-and-public-key'
+            ] + matching_keys,
+            executable=get_bin_path('gpg'),
+            check_rc=True
+        )
+        return dict(changed=True, fingerprints=matching_keys)
+    return dict(changed=False, fingerprints=[])
+
+
+def generate_keypair(module, params, matching_keys, check_mode):
+    if matching_keys and not params['force']:
+        return dict(changed=False, fingerprints=matching_keys)
+
+    parameters = '''
+        {0}
+        {1}
+        {2}
+        {3}
+        {4}
+        {5}
+        {6}
+        {7}
+        %commit
+        '''.format(
+        'Key-Type: {0}'.format(params['key_type'] if params['key_type'] else 'default'),
+        'Key-Length: {0}'.format(params['key_length']) if params['key_length'] else '',
+        'Key-Curve: {0}'.format(params['key_curve']) if params['key_curve'] else '',
+        'Expire-Date: {0}'.format(params['expire_date']) if params['expire_date'] else '',
+        'Name-Real: {0}'.format(params['name']) if params['name'] else '',
+        'Name-Comment: {0}'.format(params['comment']) if params['comment'] else '',
+        'Name-Email: {0}'.format(params['email']) if params['email'] else '',
+        'Passphrase: {0}'.format(params['passphrase']) if params['passphrase'] else '%no-protection',
+    )
+
+    rc, stdout, stderr = module.run_command(
+        [
+            '--dry-run' if check_mode else '',
+            '--batch',
+            '--gen-key',
+        ],
+        data=parameters,
+        executable=get_bin_path('gpg'),
+        check_rc=True
+    )
+
+    fingerprint = re.search(r'([0-9A-Z]+)\.rev', stderr).group(1)
+
+    for subkey in params['subkeys']:
+        if subkey['subkey_type'] in ['RSA', 'DSA', 'ELG']:
+            algo = '{0}'.format(subkey['subkey_type'].lower())
+            if subkey['subkey_length']:
+                algo += str(subkey['subkey_length'])
+        elif subkey['subkey_curve']:
+            algo = subkey['subkey_curve']
+        else:
+            algo = 'default'
+
+        module.run_command(
+            [
+                '--dry-run' if check_mode else '',
```

Comment:
I don't think providing an empty argument is a good idea. If you want to skip it, use `None`, since `AnsibleModule.run_command()` will simply omit `None` values (as opposed to empty strings).